src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -878,13 +880,36 @@ internal void InitializePropertyCache()
                 ExtensionDataProperty.EnsureConfigured();
             }
 
+            int numberOfRequiredProperties = 0;
             foreach (KeyValuePair<string, JsonPropertyInfo> jsonPropertyInfoKv in PropertyCache.List)
             {
                 JsonPropertyInfo jsonPropertyInfo = jsonPropertyInfoKv.Value;
 
+                if (jsonPropertyInfo.IsRequired)
+                {
+                    numberOfRequiredProperties++;
+                }
+
                 jsonPropertyInfo.EnsureChildOf(this);
                 jsonPropertyInfo.EnsureConfigured();
             }
+
+            if (numberOfRequiredProperties > 0)
```

Comment:
Does avoiding array resizes necessitate looping over the properties twice? I would recommend using something like
```C#
List<JsonPropertyInfo>? requiredProperties = null;

for (...)
{
    if (jsonPropertyInfo.IsRequired)
    {
        (requiredProperties ??= new()).Add(jsonPropertyInfo);
    }
}

if (requiredProperties != null)
{
    requiredProperties = requiredProperties.ToArray();
}
```

Reply:
we usually deal with relatively small amount of properties. I think it's faster this way since we avoid allocation. We iterate couple of times anyway (i.e. to create cache)

Reply:
This approach also avoids allocation unless at least one `IsRequired` property is encountered.

Reply:
I meant scenario when it's encountered we will do it twice with the suggested code

Reply:
Sure, but I don't see why this might be a problem? This concerns metadata initialisation and not the serialization hot path. I don't think it is a micro-optimization worth making our code more complicated for.